File: 2013-detect-squares/2013-detect-squares.js

var DetectSquares = function() {
    this.hash = {}
};

/** 
 * @param {number[]} point
 * @return {void}
 */
DetectSquares.prototype.add = function(point) {
    let str = point[0] + "-" + point[1]
    let map = this.hash
    map[str] = (map[str] || 0) + 1
};

/** 
 * @param {number[]} point
 * @return {number}
 */
DetectSquares.prototype.count = function(point) {
     let [x, y]  = point
    let map = this.hash
    let count = 0
    for(let key in map){
        let [x1, y1] = key.split("-")
        x1 = +x1
        y1 = +y1
        if(Math.abs(x - x1) == 0 || Math.abs(y1 - y) !== Math.abs(x1 - x)) continue
        let first = map[x + "-" + y1] || 0
        let second = map[x1 + "-" + y] || 0
        count += first * second * map[key]
    }
    
    

    return count

};

/** 
 * Your DetectSquares object will be instantiated and called as such:
 * var obj = new DetectSquares()
 * obj.add(point)
 * var param_2 = obj.count(point)
 */

File: 2013-detect-squares/2013-detect-squares.py
class DetectSquares:

    def __init__(self):
        self.pointsPerY = defaultdict(Counter)
        

    def add(self, point: List[int]) -> None:
        x, y = point
        self.pointsPerY[y][x] +=1
        

    def count(self, point: List[int]) -> int:
        
        x1 , y = point
        points = self.pointsPerY[y]
        ans = 0
        for x2 in points.keys():
            if x1 == x2 :
                continue
            ans += self.pointsPerY[y][x2] * self.pointsPerY[y-abs(x1-x2)][x2] * self.pointsPerY[y-abs(x1-x2)][x1]
            ans += self.pointsPerY[y][x2] * self.pointsPerY[y+abs(x1-x2)][x2] * self.pointsPerY[y+abs(x1-x2)][x1]
        return ans     

        
        


# Your DetectSquares object will be instantiated and called as such:
# obj = DetectSquares()
# obj.add(point)
# param_2 = obj.count(point)